src/ztex_md5crypt.c

```diff
@@ -0,0 +1,249 @@
+/*
+ * This software is Copyright (c) 2018 Denis Burykin
+ * [denis_burykin yahoo com], [denis-burykin2014 yandex ru]
+ * and it is hereby released to the general public under the following terms:
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted.
+ *
+ * This file is part of John the Ripper password cracker,
+ * Copyright (c) 1996-2001,2008,2010-2013,2015 by Solar Designer
+ *
+ * This file is part of John the Ripper password cracker,
+ * Copyright (c) 2012 magnum / JimF
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted.
+ *
+ * There's ABSOLUTELY NO WARRANTY, express or implied.
+ *
+ */
+#include <stdio.h>
+#include <string.h>
+
+#include "arch.h"
+#include "common.h"
+#include "formats.h"
+#include "memory.h"
+#include "config.h"
+#include "options.h"
+#include "MD5_std.h"
+#include "md5crypt_common.h"
+
+#include "ztex/device_bitstream.h"
+#include "ztex/device_format.h"
+#include "ztex/task.h"
+#include "ztex/pkt_comm/cmp_config.h"
+
+
+#define FORMAT_LABEL		"md5crypt-ztex"
+#define ALGORITHM_NAME		"md5crypt ZTEX"
+
+#define FORMAT_NAME		"crypt(3) $1$"
+#define FORMAT_TAG			"$1$"
+#define FORMAT_TAG_LEN		(sizeof(FORMAT_TAG)-1)
+
+#define BENCHMARK_COMMENT	""
+#define BENCHMARK_LENGTH	-1
+
+#define CIPHERTEXT_LENGTH		22
+
+// Data structures differ from ones from CPU implementation:
+// - Only partial hash in the database (BINARY_SIZE=4)
+// - The hardware allows candidate length <= 64
+//
+#define PLAINTEXT_LENGTH	64
+#define BINARY_SIZE			4
+#define BINARY_ALIGN		4
+#define SALT_SIZE			9
+#define SALT_ALIGN			1
+
+
+static struct device_bitstream bitstream = {
+	// bitstream ID (check is performed by querying operating bitstream)
+	0xd5c0,
+	"$JOHN/ztex/ztex115y_md5crypt.bit",
+	// parameters for high-speed packet communication (pkt_comm)
+	{ 2, 14336, 4094 },
+	// computing performance estimation (in candidates per interval)
+	// (keys * mask_num_cand)/crypt_all_interval per jtr_device.
+	230400,		// keys/fpga for crypt_all()
+	16384,		// keys/fpga for self-test
+	1536 * 1024,	// Would be 48 MB of USB traffic on 32-byte keys
+	512,		// Max. number of entries in onboard comparator.
+	384,		// Min. number of keys for effective device utilization
+	1, { 135 },	// Programmable clocks
```

Comment:
Should probably use same default frequency as we have in the comment in john.conf, currently 180.